packages/document-event/src/hotKey.ts

```diff
@@ -0,0 +1,154 @@
+    if (options.strict) {
+        // check both truthy and falsy
+        return (
+            Boolean(e.ctrlKey) === Boolean(options.ctrlKey)
+            && Boolean(e.shiftKey) === Boolean(options.shiftKey)
+            && Boolean(e.altKey) === Boolean(options.altKey)
+            && Boolean(e.metaKey) === Boolean(options.metaKey)
+        );
+    }
+    // not strict
+    const keys = composingKeys.filter(
+        key => options[key] !== undefined
+    );
+
+    for (const key of keys) {
+        if (e[key] !== options[key]) {
+            return false;
+        }
+    }
+    return true;
+};
+
+const isKeyMatched = (e: KeyboardEvent, options: Options) => {
+    if (e.code === options.code) {
+        return true;
+    }
+    if (e.key === options.key) {
+        return true;
+    }
+    if (options.code === undefined && options.key === undefined) {
+        return true;
+    }
+    return false;
+};
+
+/**
+ * 判断元素是否为表单元素
+ * @see https://github.com/github/hotkey/blob/main/src/utils.ts#L1
+ */
+const isFormField = (element: Node) => {
+    if (!(element instanceof HTMLElement)) {
+        return false;
+    }
+
+    const name = element.nodeName.toLowerCase();
+    const type = (element.getAttribute('type') || '').toLowerCase();
+    return (
+        name === 'select'
+        || name === 'textarea'
+        || (name === 'input' && type !== 'submit' && type !== 'reset' && type !== 'checkbox' && type !== 'radio')
+        || element.isContentEditable
+    );
+};
+
+const isMatched = (e: KeyboardEvent, options: Options) => {
+    return (
+        !isFormField(e.target as HTMLElement)
+        && isKeyMatched(e, options)
+        && isComposingMatched(e, options)
+        && isRepeatMatched(e, options)
+    );
+};
+
+const createHotKeyHook = () => {
+    const optionList: Options[] = [];
+    const onDocumentKeydown = (e: KeyboardEvent) => {
+        // 是否需要一个参数来定义是否需要 prevent
+        // e.preventDefault();
+        optionList.forEach(options => {
+            const {keydown} = options;
+            if (!keydown) {
+                return;
+            }
+            if (isMatched(e, options)) {
+                keydown(e);
+            }
+        });
+    };
+    const onDocumentKeyup = (e: KeyboardEvent) => {
+        // 是否需要一个参数来定义是否需要 prevent
+        // e.preventDefault();
+        optionList.forEach(options => {
+            const {keyup} = options;
+            if (!keyup) {
+                return;
+            }
+            /**
+             * @see https://cloud.tencent.com/developer/ask/70262
+             */
+            if (e.key === 'Meta') {
+                keyup(e);
+            } else if (isMatched(e, options)) {
+                keyup(e);
+            }
+        });
+    };
+    const useHotKey = (options: Options) => {
+        const memoizedOptions = useMemo(
+            () => options,
+            // eslint-disable-next-line react-hooks/exhaustive-deps
+            Object.values(options)
```

Comment:
This is unsafe since `Object.values` relies on the property order when object is initialized, `Object.values({x: 1, y: 2})` doesn't equals to `Object.values({y: 2, x: 1})`.

Also a "same value but different name" could break this implementation, like `{ctrl: true}` -> `{alt: true}`

I think it could be safe if we:

1. Make `optionList` a `Set`
2. Put `options` into `Set` in `useEffect` and remove it in clean up function

In this case we can keep `options` update to date and unique even its reference changes on each render

Reply:
fair